### ISSUE TITLE
"Will set the next sink as the FallbackSink" fails with "attempt to index a nil value in sink[s]"

### ISSUE THREAD

**Issue by doronbehar** [stefano-m/lua-pulseaudio_dbus#7]
Continuing from https://github.com/stefano-m/lua-pulseaudio_dbus/commit/4a62266b14a6ef007a713465ab69b37dd8eea9ae#commitcomment-25384458, I came to this thought because `busted` gives me an error in [line 175](https://github.com/stefano-m/lua-pulseaudio_dbus/blob/master/tests/pulseaudio_dbus_spec.lua#L175) in `tests/pulseaudio_dbus_spec.lua` about an attempt to index a nil value in `sink[s]`.
I can't say that using `sink` as an array of objects is the best solution but IMHO it is the most logical one.

Anyway we need to solve it in any way you choose to.
Thanks.

**Comment by stefano-m** [comment]
I have just realised that the code in that test is wrong. I had not noticed earlier because I can't manage to run it. Can you please explain how you run it so I can reproduce the problem please?

The correct way of running the test would be to use `core.get_device` to get a device object for each sink path in `core.Sinks`.

**Comment by doronbehar** [comment]
You are absolutely correct, what you said is exactly what I meant we should do.
When I run the added tests with `--filter=FallbackSink`, I have 2 sinks connected - the 1st is headphones anolog output and the 2nd is HDMI digital output, and while there is at least one sink input available.
The expected result is that after both of these tests, the sound will transition from one sink to the next one and the next sink in the available `core.Sinks` will stay as the `FallbackSink`.

**Comment by stefano-m** [comment]
Would you be willing to implement the fix using `core.get_device` then please?

I'll try and run the tests again once I have time, but I could not trigger the test when I had both HDMI and laptop speakers, so I may be doing something wrong.

Thanks.

**Comment by doronbehar** [comment]
OK, the situation might be more complicated than it seems:

You said you could not trigger the test when you had both HDMI and laptop speakers. There is a bug in PulseAudio that was fixed but not yet released (at least not in my distribution) about an issue that might be affecting you as well and [I reported it](https://bugs.freedesktop.org/show_bug.cgi?id=101794) when I noticed it.
My suspicion is that similarly to what I described in [the bug report](https://bugs.freedesktop.org/show_bug.cgi?id=101794), your `pulseaudio` daemon doesn't see all sinks available although new sinks are connected during run-time, and that's why you can't really perform the tests I added to `tests/pulseaudio_dbus_spec.lua`.

If you do experience the same problem I reported in PulseAudio's bug tracker, than you can either compile and install PulseAudio from source and restart it or temporarily make sure you have 2 sinks connected when you start `pulseaudio`. This will ensure both sinks are available for `pulseaudio` and fully detected by it.

**Comment by stefano-m** [comment]
Thanks for the information, I'll look into that.

In the meantime, would you like to fix the issue yourself and submit a PR. If so, I can assign the issue to you.

Thanks again!

**Comment by doronbehar** [comment]
Any new thoughts on this?

**Comment by stefano-m** [comment]
Hi, unfortunately I don't have time to look at this at the moment. I will get back to it at some point though.

**Comment by stefano-m** [comment]
Hi @doronbehar 
sorry for the long wait, as you may have imagined I have been quite busy.

To recap:

- Regarding https://github.com/stefano-m/lua-pulseaudio_dbus/issues/7#issuecomment-342973823 I am (now) using `pulseaudio 11.1`, I *think* that that should have the fix you mention, so I may be able to try and reproduce the issue now.

- Regarding https://github.com/stefano-m/lua-pulseaudio_dbus/issues/7#issuecomment-344973810 apologies, that should have been `pulse.get_device`, i.e. `get_device` is a top-level function: https://github.com/stefano-m/lua-pulseaudio_dbus/blob/b0f6774870658c2f776af6ad4596037f2f60191e/src/pulseaudio_dbus/init.lua#L368

Thanks.

**Comment by doronbehar** [comment]
Regarding the problem I described with upstream pulseaudio, I asked them whether the fix they implemented was released or not and the answer is **no**: https://bugs.freedesktop.org/show_bug.cgi?id=101794#c10

As for your second bullet, I'm not sure I understood what you mean in "`get_device`" (being) a top-level function". I think we are making it way too complicated and I truly think that using `sink[1]` as the main object upon which most of the single-sink tests will be executed is the most logical choice and it is best implemented in my branch - [`multiple-sinks-rerefined`](https://github.com/doronbehar/lua-pulseaudio_dbus/tree/multiple-sinks-rerefined) or in [`multiple-sinks-final`](https://github.com/doronbehar/lua-pulseaudio_dbus/tree/multiple-sinks-final) (#8) as well.

**Comment by stefano-m** [comment]
> As for your second bullet, I'm not sure I understood what you mean in "get_device" (being) a top-level function". 

I just meant that one should use `pulse.get_device` rather than `core.get_device`.

> I think we are making it way too complicated and I truly think that using sink[1]...

I disagree that pulling in all available sinks for tests that don't need them is a good thing to do. However, I can consider this if you limit the scope to just the failing test.

**Comment by doronbehar** [comment]
How about this patch for the test `"Will set the next sink as the FallbackSink"`:
```
diff --git i/tests/pulseaudio_dbus_spec.lua w/tests/pulseaudio_dbus_spec.lua
index b61ac1a..c3536aa 100644
--- i/tests/pulseaudio_dbus_spec.lua
+++ w/tests/pulseaudio_dbus_spec.lua
@@ -170,10 +170,15 @@ describe("PulseAudio with DBus", function ()
                     print("\nNOTE: Won't set the next sink as the FallbackSink because there is only one sink available in this machine")
                     return
                   end
+                  local sinks_array = {}
+                  sinks_array[1] = sink
+                  for s=2,total_number_of_sinks do
+                      sinks_array[s] = pulse.get_device(connection, assert(core.Sinks[s]))
+                  end
                   for s=1,total_number_of_sinks do
-                    if core.FallbackSink ~= sink[s].object_path then
-                      core:set_fallback_sink(sink[s].object_path)
-                      assert.is_equal(core.FallbackSink, sink[s].object_path)
+                    if core.FallbackSink ~= sinks_array[s].object_path then
+                      core:set_fallback_sink(sinks_array[s].object_path)
+                      assert.is_equal(core.FallbackSink, sinks_array[s].object_path)
                       return
                     end
                   end
```
Fully tested with 2 sinks available.

**Comment by stefano-m** [comment]
Hi, that looks good to me. If the tests pass I am OK to merge it.

Can you attach a patch prepared with `git format-patch` please?

**Comment by doronbehar** [comment]
[last-patch.txt](https://github.com/stefano-m/lua-pulseaudio_dbus/files/1841074/last-patch.txt)

**Comment by stefano-m** [comment]
Thanks!

Closed by 9a5aa75
